service/controller/clusterapi/v17/resources/awsclusterconfig/create.go

```diff
@@ -21,7 +28,130 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		return nil
 	}
 
-	// TODO: Map Cluster -> AWSClusterConfig.
+	// ClusterID is core part of e.g. PKI initialization etc. so it must be
+	// present before proceeding further.
+	if key.ClusterID(cluster) == "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("provider status in cluster cr %q does not contain cluster ID", cluster.Name))
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	var versionBundles []versionbundle.Bundle
+	{
+		req := searcher.Request{
+			ReleaseVersion: key.ClusterReleaseVersion(cluster),
+		}
+
+		res, err := r.clusterClient.Release.Searcher.Search(ctx, req)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		versionBundles = res.VersionBundles
+	}
+
+	// Get existing AWSClusterConfig or create a new one.
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if AWSClusterConfig %q/%q exists", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+
+	presentAWSClusterConfig, err := r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Get(key.AWSClusterConfigName(cluster), metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find AWSClusterConfig %q/%q", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating AWSClusterConfig %q/%q", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+
+		newAWSClusterConfig := r.constructAWSClusterConfig(cluster, versionBundles)
+
+		_, err = r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Create(&newAWSClusterConfig)
+		if errors.IsAlreadyExists(err) {
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("AWSClusterConfig %q/%q already exists", newAWSClusterConfig.Namespace, newAWSClusterConfig.Name))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created AWSClusterConfig %q/%q", newAWSClusterConfig.Namespace, newAWSClusterConfig.Name))
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found AWSClusterConfig %q/%q", cluster.Namespace, presentAWSClusterConfig.Name))
```

Comment:
I think this log should be in final `} else {` block.

Reply:
Why? The whole rest of the function is that else block.

Reply:
I missed `return nil`. Then we miss `r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")`. I'll add a suggestion.